### PR TITLE
Thread event register with usr data

### PR DESCRIPTION
**Pull request by johnfxgalea**
Enables the passing of user data for callback functions associated with init thread events.

Reference to #2826

Once you are okay with this implementation, I will implement the same for exit thread events, pushing to this branch (so don't merge as of yet?).

I would also like to enquire what is the best way to have tests for this functionality.

### CODE REVIEW

File: ext/drmgr/drmgr.c [hunk 1]
```diff
@@ -1256,7 +1263,16 @@ drmgr_register_thread_init_event_ex(void (*func)(void *drcontext),
                                     drmgr_priority_t *priority)
 {
     return drmgr_generic_event_add(&cb_list_thread_init, thread_event_lock,
-                                   (void (*)(void)) func, priority);
+                                   (void (*)(void)) func, priority, NULL);
+}
+
+DR_EXPORT
+bool
+drmgr_register_thread_init_event_usr_data(void (*func)(void *drcontext, void *usr_data),
```
Comment (derekbruening) [inline ext/drmgr/drmgr.c]: I think including the e (i.e., ...event_user_data) is much nicer in the function name.  I would also vote for spelling it out everywhere else.  WDYT?

File: ext/drmgr/drmgr.c [hunk 2]
```diff
@@ -1718,7 +1757,12 @@ drmgr_thread_init_event(void *drcontext)
     for (i = 0; i < iter.num; i++) {
         if (!iter.cbs.generic[i].pri.valid)
             continue;
-        (*iter.cbs.generic[i].cb.thread_cb)(drcontext);
+        void *usr_data = iter.cbs.generic[i].usr_data;
+        if (usr_data == NULL)
+            (*iter.cbs.generic[i].cb.thread_cb.cb_no_usr_data)(drcontext);
```
Comment (derekbruening) [inline ext/drmgr/drmgr.c]: Hmm, this seems unsafe: if I call drmgr_register_thread_init_event_usr_data(myfunc, NULL) and then in myfunc I check the value of the user_data param I'll be looking at an uninitialized value.

File: ext/drmgr/drmgr.c [hunk 3]
```diff
@@ -1739,7 +1783,11 @@ drmgr_thread_exit_event(void *drcontext)
...
```
Comment (derekbruening) [inline ext/drmgr/drmgr.c]: Ditto.

File: ext/drmgr/drmgr.h [hunk 4]
```diff
@@ -785,6 +806,17 @@ bool
...
+DR_EXPORT
+/**
+ * Registers a callback function for the thread exit event,
+ * ordered by \p priority. Allows for the passing of user data \p usr_data
+ * which is available upon the execution of the callback. drmgr calls \p func
+ * when an application thread exits. \return whether successful.
+ */
+bool
+drmgr_register_thread_exit_event_usr_data(void (*func)(void *drcontext, void *usr_data),
+                                    drmgr_priority_t *priority, void *usr_data);
```
Comment (derekbruening) [inline ext/drmgr/drmgr.h]: indentation is off

File: ext/drmgr/drmgr.h [hunk 5]
```diff
@@ -756,6 +756,17 @@ bool
...
 
+DR_EXPORT
+/**
+ * Registers a callback function for the thread initialization event,
+ * ordered by \p priority. Allows for the passing of user data \p usr_data
+ * which is available upon the execution of the callback. drmgr calls \p func
+ * whenever the application creates a new thread. \return whether successful.
+ */
+bool
+drmgr_register_thread_init_event_usr_data(void (*func)(void *drcontext, void *usr_data),
```
Comment (derekbruening) [inline ext/drmgr/drmgr.h]: Please add a test to suite/tests/client-interface/drmgr-test.dll.c

Comment (derekbruening) [inline ext/drmgr/drmgr.h]: Please add a line in api/docs/release.dox (at the bottom of the list [here](https://github.com/DynamoRIO/dynamorio/blob/master/api/docs/release.dox#L230))

File: ext/drmgr/drmgr.h [hunk 6]
```diff
@@ -785,6 +806,17 @@ bool
...
 
+DR_EXPORT
+/**
+ * Registers a callback function for the thread exit event,
+ * ordered by \p priority. Allows for the passing of user data \p usr_data
+ * which is available upon the execution of the callback. drmgr calls \p func
+ * when an application thread exits. \return whether successful.
+ */
+bool
+drmgr_register_thread_exit_event_usr_data(void (*func)(void *drcontext, void *usr_data),
```
Comment (derekbruening) [inline ext/drmgr/drmgr.h]: Ditto on the test and release.dox feature line (I'd combine with the init one in one line)

**Comment by johnfxgalea** [comment]
I think it's ready to be checked. I already have a similar implementation for module load/unload events.

File: suite/tests/client-interface/drmgr-test.dll.c [hunk 7]
```diff
@@ -121,6 +133,12 @@ dr_init(client_id_t id)
                                   NULL, NULL, 10};
     drmgr_priority_t sys_pri_B = {sizeof(priority), "drmgr-test-B",
                                   "drmgr-test-A", NULL, 5};
+    drmgr_priority_t thread_init_null_user_data_pri = {sizeof(priority),
+                                               "drmgr-thread-init-null-user-data-test",
+                                               NULL, NULL, -3};
```
Comment (derekbruening) [inline suite/tests/client-interface/drmgr-test.dll.c]: Indentation looks off on these 2 lines

File: suite/tests/client-interface/drmgr-test.dll.c [hunk 8]
```diff
@@ -112,6 +122,8 @@ static dr_emit_flags_t one_time_bb_event(void *drcontext, void *tag, instrlist_t
                                          bool for_trace, bool translating);
 static void event_kernel_xfer(void *drcontext, const dr_kernel_xfer_info_t *info);
 
+uintptr_t user_data_test = 9090;
```
Comment (derekbruening) [inline suite/tests/client-interface/drmgr-test.dll.c]: const and static

**Comment by derekbruening** [comment]
run aarch64 tests

**Comment by derekbruening** [comment]
I filed https://github.com/DynamoRIO/dynamorio/issues/2833 to cover this general feature: please use it for future PR's.

**Pull request merged**